include/fi_perf.h

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (c) 2015-2017 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_PERF_H_
+#define _FI_PERF_H_
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+#include <stdio.h>
+
+/* Generic logic */
+
+#define FI_PERF_SLOT_COUNT_MAX       (8)
+#define FI_PERF_SLOT_NAME_LENGTH_MAX (256)
+#define FI_PERF_PRIVATE_AREA_SIZE    (64)
+
+struct fi_perf_slot {
+	uint64_t begin;
+	uint64_t sum;
+	uint64_t count;
+	uint64_t is_active;
+};
+
+struct fi_perf_handle {
+	struct fi_perf_slot slots[FI_PERF_SLOT_COUNT_MAX];
+	char slot_names[FI_PERF_SLOT_COUNT_MAX][FI_PERF_SLOT_NAME_LENGTH_MAX];
+	int count;
+	uint8_t private[FI_PERF_PRIVATE_AREA_SIZE];
+};
+
+#ifdef FI_PERF_RDPMC
```

Comment:
I'd prefer a more generic name, versus referring to an intel instruction.  E.g. maybe FI_PERF_PROFILE or FI_PERF_ENABLED?  Or are we really limited to only intel here?  I'm okay with the implementation being limited, but the architecture should strive to be generic.

Reply:
The code is RDPMC instruction specific. The idea is to have a generic lightweight API which would work with slots/regions (could be associated with function or any arbitrary code region) and begin/end markers. It should be possible to have different back ends. RDPMC based approach is one of the instances. (I may add RDTSC based implementation as well)

Too generic approach may lead to a serious overhead so we have to find a good balance. The most common scenario we have is 1-2 slots/regions and only one type of HW event. In that case it is OK to have the markers on a critical path w/o significant impact on analyzed scenario.
E.g. PAPI (http://icl.utk.edu/papi/) is pretty heavy in some cases.

Regarding your idea about profiling provider we may wrap OFI functions, set number of slots equal to number of profiled functions and each function will have begin/end in its decorator.